src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -0,0 +1,144 @@
+#include "cluster_task_manager.h"
+
+#include "ray/util/logging.h"
+
+namespace ray {
+namespace raylet {
+
+ClusterTaskManager::ClusterTaskManager(
+    const ClientID &self_node_id,
+    std::shared_ptr<ClusterResourceScheduler> cluster_resource_scheduler,
+    std::function<bool(const Task &)> fulfills_dependencies_func,
+    std::shared_ptr<gcs::GcsClient> gcs_client)
+    : self_node_id_(self_node_id),
+      cluster_resource_scheduler_(cluster_resource_scheduler),
+      fulfills_dependencies_func_(fulfills_dependencies_func),
+      gcs_client_(gcs_client) {}
+
+bool ClusterTaskManager::SchedulePendingTasks() {
+  size_t queue_size = tasks_to_schedule_.size();
+  bool did_schedule = false;
+
+  // Check every task in task_to_schedule queue to see
+  // whether it can be scheduled. This avoids head-of-line
+  // blocking where a task which cannot be scheduled because
+  // there are not enough available resources blocks other
+  // tasks from being scheduled.
+  while (queue_size-- > 0) {
+    Work work = tasks_to_schedule_.front();
+    tasks_to_schedule_.pop_front();
+    Task task = work.second;
+    auto request_resources =
+        task.GetTaskSpecification().GetRequiredResources().GetResourceMap();
+    int64_t violations = 0;
```

Comment:
Does this value get used?

Reply:
It is used if a resource request is soft. (Out of scope at the moment, but probably not worth removing since we'll re-add it soon).